base/learner.h

```diff
@@ -0,0 +1,45 @@
+#ifndef  _MACHINELEARNING_BASE_LEARNER_H_ 
+#define  _MACHINELEARNING_BASE_LEARNER_H_
+
+#include <vector>
+
+namespace base {
+
+/**
+ * Base class for all learning algorithm implementations.
+ */
+class Learner{
+
+  public:
+
+    /**
+     * \brief Standard destructor.
+     */
+    virtual ~Learner();
+
+    /**
+     * Train the learner with a given dataset.
+     * \param X A pointer object of Collection.
+     */
+    virtual void Train(/*const Collection *X*/);
+
+    /**
+     * Predicts the labels of the instances of X based on the fitted model.
+     * \param X           A pointer object of Collection.
+     * \param prediction  Instances prediction.
+     */
+    virtual void Predict(/*const Collection *X,*/ std::vector<double> &prediction) = 0;
```

Comment:
As spoken with @raphaelcampos, I'm not sure how to handle the vector of predictions. I think that it is a better idea to just return a vector with the values instead of passing out through parameters. And in fact, I'm not sure if the way of doing this is by just returning a pointer or just a copy of this vector. Wrapping up, my idea is to replace the line:
`virtual void Predict(/*const Collection *X,*/ std::vector<double> &prediction) = 0;`
For:
`virtual std::vector<double> * Predict(/*const Collection *X*/) = 0;
or even do not return the pointer at all:
`virtual std::vector<double> Predict(/*const Collection *X*/) = 0;

What do you guys think?
@thiagovas @gcmgomes @luizfgoncalves